File: day16b.py
#!/usr/bin/env python3

import fileinput
import itertools
import numpy as np
import os
import sys
from pprint import pprint

DEBUG = os.environ.get("DEBUG")
PROGRESS = os.environ.get("PROGRESS")

def dprint(*args, **kwargs):
    if DEBUG:
        print(*args, **kwargs, file=sys.stderr)

REPEATS = int(os.environ.get("REPEATS", 10000))
OFFSET_DIGITS = int(os.environ.get("OFFSET_DIGITS", 7))
PHASES = int(os.environ.get("PHASES", 100))

base_pattern = [0, 1, 0, -1]

def main():
    f = fileinput.input()
    input_line = next(f).rstrip()
    input_digits = [int(c) for c in input_line]

    signal = np.tile(np.array(input_digits), REPEATS)
    n = len(signal)
    for phase in range(PHASES):
        cum = np.cumsum(signal)
        out_signal = np.zeros((n,), dtype=int)
        for out_index in range(n):
            if PROGRESS and out_index % 10000 == 0:
                sys.stderr.write(".")
                sys.stderr.flush()

            block_size = out_index + 1
            pattern_index = 0
            i = -1 # unshift
            s = 0
            while i < n:
                c = base_pattern[pattern_index]
                if c != 0:
                    if block_size == 1:
                        bs = signal[i]
                    else:
                        j = min(n - 1, i + block_size - 1)
                        bs = cum[j] - cum[i - 1]
                    s += c * bs
                i += block_size
                pattern_index = (pattern_index + 1) % len(base_pattern)

            out_signal[out_index] = abs(s) % 10
        if PROGRESS:
            print(file=sys.stderr)

        signal = out_signal

    if OFFSET_DIGITS == 0:
        offset = 0
    else:
        offset = int(input_line[:OFFSET_DIGITS], 10)

    message = signal[offset:offset + 8]

    print(''.join(map(str, message)))

main()

# vim: set shiftwidth=4 tabstop=4 :

File: day16b.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

var basePattern = []int64{0, 1, 0, -1}

func main() {
	inputLine := strings.TrimSpace(<-inputLines())
	inputDigits := []int64{}
	for i := 0; i < len(inputLine); i++ {
		x, _ := strconv.Atoi(inputLine[i:i+1])
		inputDigits = append(inputDigits, int64(x))
	}
	m := len(inputDigits)

	repeats := 10000

	signal := make([]int64, len(inputDigits) * repeats)
	for i := 0; i < repeats; i++ {
		copy(signal[i * m:(i+1)*m], inputDigits)
	}
	n := len(signal)

	for phase := 0; phase < 100; phase++ {
		cum := make([]int64, len(signal))
		cum[0] = signal[0]
		for i := 1; i < len(signal); i++ {
			cum[i] = cum[i-1] + signal[i]
		}

		outSignal := make([]int64, n)
		for outIndex := 0; outIndex < n; outIndex++ {
			blockSize := outIndex + 1
			patternIndex := 0
			i := -1
			s := int64(0)
			for i < n {
				c := basePattern[patternIndex]
				if c != 0 {
					var blockSum int64
					if blockSize == 1 {
						blockSum = signal[i]
					} else {
						j := min(n - 1, i + blockSize - 1)
						blockSum = cum[j] - cum[i - 1]
					}
					s += c * blockSum
				}
				i += blockSize
				patternIndex = (patternIndex + 1) % len(basePattern)
			}
			outSignal[outIndex] = abs(s) % 10
		}

		signal = outSignal
	}

	offset, _ := strconv.Atoi(inputLine[:7])
	message := signal[offset:offset+8]
	for _, d := range message {
		fmt.Print(d)
	}
	fmt.Println()
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func abs(a int64) int64 {
	if a < 0 {
		return -a
	} 
	return a
}

func inputLines() <-chan string {
	ch := make(chan string)

	go func() {
		inFile, err := os.Open(os.Args[1])
		if err != nil {
			log.Fatal(err)
		}

		scanner := bufio.NewScanner(inFile)
		for scanner.Scan() {
			line := scanner.Text()
			ch <- line
		}
		if err := scanner.Err(); err != nil {
			log.Fatal(err)
		}

		close(ch)
	}()

	return ch
}
